contracts/lens/QuoterV3.sol

```diff
@@ -0,0 +1,232 @@
+            uint160 sqrtPriceX96After,
+            uint32 initializedTicksCrossed,
+            uint256 gasEstimate
+        )
+    {
+        bool zeroForOne = params.tokenIn < params.tokenOut;
+        IUniswapV3Pool pool = getPool(params.tokenIn, params.tokenOut, params.fee);
+
+        uint256 gasBefore = gasleft();
+        try
+            pool.swap(
+                address(this), // address(0) might cause issues with some tokens
+                zeroForOne,
+                params.amountIn.toInt256(),
+                params.sqrtPriceLimitX96 == 0
+                    ? (zeroForOne ? TickMath.MIN_SQRT_RATIO + 1 : TickMath.MAX_SQRT_RATIO - 1)
+                    : params.sqrtPriceLimitX96,
+                abi.encodePacked(params.tokenIn, params.fee, params.tokenOut)
+            )
+        {} catch (bytes memory reason) {
+            gasEstimate = gasBefore - gasleft();
+            return handleRevert(reason, pool, gasEstimate);
+        }
+    }
+
+    /// @dev Fetch an exactIn quote for a V2 pair on chain
+    function quoteExactInputSingleV2(
+        uint256 amountIn,
+        address tokenIn,
+        address tokenOut
+    ) public view returns (uint256 amountOut) {
+        amountOut = getPairAmountOut(amountIn, tokenIn, tokenOut);
+    }
+
+    /// @dev Get the quote for an exactIn swap between an array of V2 and/or V3 pools
+    function quoteExactInput(bytes memory path, uint256 amountIn)
+        public
+        override
+        returns (
+            uint256 amountOut,
+            uint160[] memory sqrtPriceX96AfterList,
+            uint32[] memory initializedTicksCrossedList,
+            uint256 gasEstimate
+        )
+    {
+        sqrtPriceX96AfterList = new uint160[](path.numPools());
+        initializedTicksCrossedList = new uint32[](path.numPools());
+
+        uint256 i = 0;
+        while (true) {
+            (address tokenIn, address tokenOut, uint24 fee) = path.decodeFirstPool();
+
+            if (fee & flagBitmask != 0) {
+                amountIn = quoteExactInputSingleV2(amountIn, tokenIn, tokenOut);
+            } else {
+                /// the outputs of prior swaps become the inputs to subsequent ones
+                (
+                    uint256 _amountOut,
+                    uint160 _sqrtPriceX96After,
+                    uint32 _initializedTicksCrossed,
+                    uint256 _gasEstimate
+                ) =
+                    quoteExactInputSingle(
+                        QuoteExactInputSingleParams({
+                            tokenIn: tokenIn,
+                            tokenOut: tokenOut,
+                            fee: fee,
+                            amountIn: amountIn,
+                            sqrtPriceLimitX96: 0
+                        })
+                    );
+                sqrtPriceX96AfterList[i] = _sqrtPriceX96After;
+                initializedTicksCrossedList[i] = _initializedTicksCrossed;
+                gasEstimate += _gasEstimate;
```

Comment:
Yeah, unfortunately the logic doesn't work for the v2 swaps since it's just calling a view function so it'll always be like 26_000 gas. We can rename it, or optionally choose to keep it and assign it 0 since I don't think we use it anyways.

Reply:
Adding v3 prefix to sqrtPrice, init ticks, and gas estimate in latest commit